core/service/src/cryptography/encryption.rs

```diff
@@ -0,0 +1,641 @@
+use crate::{
+    consts::SAFE_SER_SIZE_LIMIT,
+    cryptography::{
+        error::CryptographyError, hybrid_ml_kem, hybrid_ml_kem::HybridKemCt,
+        internal_crypto_types::LegacySerialization,
+    },
+};
+use ml_kem::EncodedSizeUser;
+use ml_kem::KemCore;
+use ml_kem::{MlKem1024, MlKem512};
+use nom::AsBytes;
+use rand::{CryptoRng, RngCore};
+use serde::{de::Visitor, Deserialize, Deserializer, Serialize};
+use strum_macros::Display;
+use tfhe::{named::Named, safe_serialization::safe_deserialize, Versionize};
+use tfhe_versionable::VersionsDispatch;
+use zeroize::Zeroize;
+
+#[derive(Clone, Debug, Serialize, Deserialize, VersionsDispatch)]
+pub enum UnifiedPublicEncKeyVersioned {
+    V0(UnifiedPublicEncKey),
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize, Versionize)]
+#[versionize(UnifiedPublicEncKeyVersioned)]
+#[expect(clippy::large_enum_variant)]
+pub enum UnifiedPublicEncKey {
+    MlKem512(PublicEncKey<ml_kem::MlKem512>),
+    // LEGACY: Note that this should ONLY be used for legacy reasons, new code should use MlKem512.
+    // If used in current code, then take care to NOT use to_legacy_bytes or from_legacy_bytes on this variant
+    // as this will do bincode serialization instead of safe serialization
+    MlKem1024(PublicEncKey<ml_kem::MlKem1024>),
```

Comment:
maybe use `#[deprecated]` so that the compiler warns us if it's used by accident (but I'm not sure if this works on variants)

Reply:
Done (among other things) in 3cb0bf98d2bc8521fdecd54ece2477c03ff18dac